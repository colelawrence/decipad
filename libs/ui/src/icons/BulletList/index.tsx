import { FC } from 'react';
import { cssVar } from '../../primitives';

export const BulletList = (): ReturnType<FC> => (
  <svg viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
    <line
      x1="3.45005"
      y1="8.15002"
      x2="3.75005"
      y2="8.15002"
      stroke={cssVar('iconColorHeavy')}
      strokeWidth="1.3"
      strokeLinecap="round"
      strokeLinejoin="round"
    />
    <line
      x1="3.45005"
      y1="4.95"
      x2="3.75005"
      y2="4.95"
      stroke={cssVar('iconColorHeavy')}
      strokeWidth="1.3"
      strokeLinecap="round"
      strokeLinejoin="round"
    />
    <line
      x1="3.45005"
      y1="11.35"
      x2="3.75005"
      y2="11.35"
      stroke={cssVar('iconColorHeavy')}
      strokeWidth="1.3"
      strokeLinecap="round"
      strokeLinejoin="round"
    />
    <line
      x1="6.65"
      y1="8.15002"
      x2="12.55"
      y2="8.15002"
      stroke={cssVar('iconColorHeavy')}
      strokeWidth="1.3"
      strokeLinecap="round"
      strokeLinejoin="round"
    />
    <line
      x1="6.65"
      y1="4.95003"
      x2="12.55"
      y2="4.95003"
      stroke={cssVar('iconColorHeavy')}
      strokeWidth="1.3"
      strokeLinecap="round"
      strokeLinejoin="round"
    />
    <line
      x1="6.65"
      y1="11.35"
      x2="12.55"
      y2="11.35"
      stroke={cssVar('iconColorHeavy')}
      strokeWidth="1.3"
      strokeLinecap="round"
      strokeLinejoin="round"
    />
  </svg>
);
