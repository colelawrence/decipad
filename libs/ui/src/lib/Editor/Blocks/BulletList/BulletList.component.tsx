import { UnorderedList } from '@chakra-ui/react';
import React from 'react';
import { RenderElementProps } from 'slate-react';

export const BulletList = ({
  attributes,
  children,
}: RenderElementProps): JSX.Element => {
  return <UnorderedList {...attributes}>{children}</UnorderedList>;
};
