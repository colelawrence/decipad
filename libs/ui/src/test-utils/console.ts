export const mockConsoleWarn = () => {
  const originalConsoleWarn = console.warn;
  const mockConsoleWarn: jest.MockedFunction<typeof console.warn> = jest.fn();

  beforeEach(() => {
    console.warn = jest.fn();
    mockConsoleWarn.mockClear();
  });
  afterEach(() => {
    console.warn = originalConsoleWarn;
  });

  return mockConsoleWarn;
};
