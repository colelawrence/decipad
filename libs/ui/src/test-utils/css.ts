import cssVarsPonyfill from 'css-vars-ponyfill';

export const findParentWithStyle = <P extends keyof CSSStyleDeclaration>(
  element: Element | null,
  propertyName: P
):
  | (Pick<CSSStyleDeclaration, P> & {
      element: Element;
      styles: CSSStyleDeclaration;
    })
  | null => {
  if (element === null) {
    return null;
  }
  const styles = getComputedStyle(element);
  if (!styles[propertyName]) {
    return findParentWithStyle(element.parentElement, propertyName);
  }
  return {
    [propertyName]: styles[propertyName],
    element,
    styles,
    // cannot type dynamic property key based on type parameter
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any;
};

export const applyCssVars = (): Promise<() => void> => {
  return new Promise((resolve, reject) => {
    const cleanup = () =>
      [...document.head.querySelectorAll('[data-cssvars="out"]')].forEach(
        (element) => element.remove()
      );
    cssVarsPonyfill({
      onError: reject,
      onComplete: () => resolve(cleanup),
    });
  });
};

export const applyPrefersColorScheme = async (
  colorScheme: 'dark' | 'light'
): Promise<() => void> => {
  // JSDOM ignores CSSOM modifications, so build a CSS string and create and append a style element instead
  const styleText = [...document.styleSheets]
    .flatMap((styleSheet) => [...styleSheet.cssRules])
    .flatMap((rule) => {
      if (rule.type === CSSRule.MEDIA_RULE) {
        const mediaRule = rule as CSSMediaRule;
        for (
          let mediaIndex = 0;
          mediaIndex < mediaRule.media.length;
          mediaIndex += 1
        ) {
          if (
            mediaRule.media[mediaIndex].replace(/\s/g, '') ===
            `(prefers-color-scheme:${colorScheme})`
          ) {
            return [...mediaRule.cssRules];
          }
        }
      }
      return [];
    })
    .flatMap((rule) => rule.cssText)
    .join(' ');

  const styleElement = document.createElement('style');
  styleElement.innerHTML = styleText;
  document.head.appendChild(styleElement);

  const cleanupCssVars = await applyCssVars();

  return () => {
    cleanupCssVars();
    styleElement.remove();
  };
};
