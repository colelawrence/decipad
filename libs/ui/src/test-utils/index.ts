export * from './menu';
export * from './language';
