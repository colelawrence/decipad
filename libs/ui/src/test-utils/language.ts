import {
  Result,
  runCode as run,
  SerializedTypeKind,
  serializeResult,
} from '@decipad/computer';

export async function runCode<T extends SerializedTypeKind>(
  code: string
): Promise<Result.Result<T>> {
  const result = await run(code);
  return serializeResult<T>(result.type, result.value);
}
