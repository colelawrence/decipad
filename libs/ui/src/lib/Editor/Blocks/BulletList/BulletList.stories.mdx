import { Meta, Story, Canvas, Props } from '@storybook/addon-docs/blocks';
import { BulletList } from './BulletList.component';
import { ListItem } from '@chakra-ui/react';

<Meta
  title="Legacy/Editor/Blocks/Lists/BulletList"
  parameters={{
    docs: {
      source: {
        code: `<BulletList
  element={{
    children: [{ text: 'The quick brown fox jumps over the lazy dog.' }],
  }}
  attributes={{ 'data-slate-node': 'element', ref: null }}
>
  <ListItem>The quick brown fox jumps over the lazy dog.</ListItem>
</BulletList>`,
      },
    },
  }}
  argTypes={{
    children: {
      defaultValue: 'The quick brown fox jumps over the lazy dog.',
      description: 'The list items of the component',
      type: 'string',
    },
  }}
  component={BulletList}
/>

# BulletList

The BulletList component serves as one of the editor's slate blocks which is basically just a `ul` tag component that is wrapped around `li` items.

export const Template = (args) => (
  <BulletList {...args}>
    <ListItem>{args.children}</ListItem>
    <ListItem>{args.children}</ListItem>
    <ListItem>{args.children}</ListItem>
    <ListItem>{args.children}</ListItem>
    <ListItem>{args.children}</ListItem>
  </BulletList>
);

<Canvas>
  <Story name="BulletList">{Template.bind({})}</Story>
</Canvas>

<Props story="BulletList" />
