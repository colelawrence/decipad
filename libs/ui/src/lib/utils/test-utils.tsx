import { ChakraProvider } from '@chakra-ui/react';
import { render, RenderOptions } from '@testing-library/react';
import React from 'react';

import { Type } from '@decipad/language';
import { theme, ResultsContextProvider } from '../..';

const results = {
  test: {
    type: Type.build({
      type: 'number',
      unit: {
        unit: 'apples',
      },
    } as any),
    value: [20],
    errors: [],
  },
  noResult: null,
};

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <ChakraProvider resetCSS theme={theme}>
      <ResultsContextProvider value={results}>
        {children}
      </ResultsContextProvider>
    </ChakraProvider>
  );
};

const customRender = (
  ui: JSX.Element,
  options?: Omit<RenderOptions, 'queries'>
) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
