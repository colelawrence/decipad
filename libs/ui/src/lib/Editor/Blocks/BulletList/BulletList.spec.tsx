import { ListItem } from '@chakra-ui/react';
import React from 'react';
import { render } from 'test-utils';
import { BulletList } from './BulletList.component';

describe('BulletList Block', () => {
  it('renders', () => {
    const { container } = render(
      <BulletList
        element={{
          children: [{ text: 'The quick brown fox jumps over the lazy dog.' }],
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
      >
        <ListItem>The quick brown fox jumps over the lazy dog.</ListItem>
      </BulletList>
    );

    expect(container).toBeDefined();
  });

  it('matches snapshot', () => {
    const { container } = render(
      <BulletList
        element={{
          children: [{ text: 'The quick brown fox jumps over the lazy dog.' }],
        }}
        attributes={{ 'data-slate-node': 'element', ref: null }}
      >
        <ListItem>The quick brown fox jumps over the lazy dog.</ListItem>
      </BulletList>
    );

    expect(container).toMatchSnapshot();
  });
});
