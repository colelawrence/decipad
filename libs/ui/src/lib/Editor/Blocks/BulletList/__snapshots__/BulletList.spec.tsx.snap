// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulletList Block matches snapshot 1`] = `
.emotion-0 {
  list-style-type: initial;
  -webkit-margin-start: 1em;
  margin-inline-start: 1em;
}

<div>
  <ul
    class="emotion-0"
    data-slate-node="element"
    role="list"
  >
    <li
      class="emotion-1"
    >
      The quick brown fox jumps over the lazy dog.
    </li>
  </ul>
</div>
`;
